### PR TITLE
Run GoCD Docker Containers as non root user

### PR DESCRIPTION
**Pull request by GaneshSPatil**
* Add go user to root group to allow GoCD running on openshift
* Add go user in dockerfile
* Change file permissions of files used by go user to go:root

### CODE REVIEW

File: buildSrc/src/main/groovy/com/thoughtworks/go/build/docker/BuildDockerImageTask.groovy [hunk 1]
```diff
@@ -201,19 +201,6 @@ class BuildDockerImageTask extends DefaultTask {
   }
 
   Map<String, Map<String, String>> getAdditionalFiles() {
-    return [
-      '/usr/local/sbin/tini': [
```
Comment (GaneshSPatil) [inline buildSrc/src/main/groovy/com/thoughtworks/go/build/docker/BuildDockerImageTask.groovy]: bring back tini to keep track of forked child processes.

File: buildSrc/src/main/resources/gocd-docker-agent/Dockerfile.agent.ftl [hunk 2]
```diff
@@ -74,15 +75,24 @@ RUN \
...
 
 <#if distro.name() == "docker">
-ADD run-docker-daemon.sh /
+  ADD run-docker-daemon.sh /
+  RUN chown go:root /run-docker-daemon.sh \
+      && chmod -R g=u /run-docker-daemon.sh
 </#if>
-ADD docker-entrypoint.sh /
+
+USER go
```
Comment (GaneshSPatil) [inline buildSrc/src/main/resources/gocd-docker-agent/Dockerfile.agent.ftl]: move this as the last line in dockerfile.

File: buildSrc/src/main/resources/gocd-docker-server/docker-entrypoint.sh [hunk 3]
```diff
@@ -82,12 +64,9 @@ if [ "$1" = '/go-server/server.sh' ]; then
...
-
-    try exec /usr/local/sbin/tini -- /usr/local/sbin/gosu go "$0" "$@"
-  fi
 fi
 
 # these 3 vars are used by `/go-server/server.sh`, so we export
 export GO_SERVER_SYSTEM_PROPERTIES="${GO_SERVER_SYSTEM_PROPERTIES}${GO_SERVER_SYSTEM_PROPERTIES:+ }-Dgo.console.stdout=true"
 
-try exec "$@"
+try /usr/local/sbin/tini exec "$@"
```
Comment (adityasood) [inline buildSrc/src/main/resources/gocd-docker-server/docker-entrypoint.sh]: we should be executing tini with args. Current usage is incorrect and gives the following error 

```╰─$ docker run -it docker-gocd-server:v19.5.0-9192                                                                                                                                        [19/06/5| 1:55PM]
/docker-entrypoint.sh: Creating directories and symlinks to hold GoCD configuration, data, and logs
$ mkdir -v /godata/artifacts
created directory: '/godata/artifacts'
$ ln -sv /godata/artifacts /go-working-dir/artifacts
'/go-working-dir/artifacts' -> '/godata/artifacts'
$ mkdir -v /godata/config
created directory: '/godata/config'
$ ln -sv /godata/config /go-working-dir/config
'/go-working-dir/config' -> '/godata/config'
$ mkdir -v /godata/db
created directory: '/godata/db'
$ ln -sv /godata/db /go-working-dir/db
'/go-working-dir/db' -> '/godata/db'
$ mkdir -v /godata/logs
created directory: '/godata/logs'
$ ln -sv /godata/logs /go-working-dir/logs
'/go-working-dir/logs' -> '/godata/logs'
$ mkdir -v /godata/plugins
created directory: '/godata/plugins'
$ ln -sv /godata/plugins /go-working-dir/plugins
'/go-working-dir/plugins' -> '/godata/plugins'
$ mkdir -v /godata/addons
created directory: '/godata/addons'
$ ln -sv /godata/addons /go-working-dir/addons
'/go-working-dir/addons' -> '/godata/addons'
$ cp -rfv /go-server/config/logback-include.xml /go-working-dir/config/logback-include.xml
'/go-server/config/logback-include.xml' -> '/go-working-dir/config/logback-include.xml'
$ install-gocd-plugins
$ mkdir -p /godata/plugins/external
$ git-clone-config
$ cd /godata/config
/docker-entrypoint.sh: Running custom scripts in /docker-entrypoint.d/ ...
$ /usr/local/sbin/tini exec /go-server/server.sh
[WARN  tini (25)] Tini is not running as PID 1 and isn't registered as a child subreaper.
Zombie processes will not be re-parented to Tini, so zombie reaping won't work.
To fix the problem, use the -s option or set the environment variable TINI_SUBREAPER to register Tini as a child subreaper, or run Tini as PID 1.
[FATAL tini (26)] exec exec failed: No such file or directory
/docker-entrypoint.sh: cannot /usr/local/sbin/tini exec /go-server/server.sh
```


change usage to 

`try exec /usr/local/sbin/tini "$@"`

**Comment by GaneshSPatil** [comment]
@adityasood -- can this PR be merged?

**Comment by adityasood** [comment]
@GaneshSPatil you can merge this and I will verify it from master

**Comment by GaneshSPatil** [comment]
The PR build VSM: https://build.gocd.org/go/pipelines/value_stream_map/trigger/4802
Installers for testing can be found: https://build.gocd.org/go/pipelines/installers-PR/1610/docker/1

**Comment by arvindsv** [comment]
@GaneshSPatil Can you write this down somewhere?

1. What happens to someone who was using an older version of the GoCD docker image? Consider cases where they had done a `chown 1000:1000` on their shared volume, as well.

2. What happens to someone using these images with Docker elastic agents? With Kubernetes elastic agents (not in Openshift)?

3. What changes need to be done, if any, by anyone using one of these images in the `FROM` part of a Dockerfile?

**Comment by ketan** [comment]
I was planning on adding a comment to some issue/PR detailing some of the
installer changes, and link that from the releases page.

Perhaps we could do something similar here, unless someone has a better
idea.

**Comment by arvindsv** [comment]
If someone needs to be added to the documentation, feel free to do that too.

**Comment by GaneshSPatil** [comment]
@arvindsv @ketan -- We were wondering if we could document this changeset as part of a blog post.

[Here](https://gist.github.com/GaneshSPatil/dabf81c95b440b77776930bbb4571ecc)  is a gist of the blog post, the content requires a bit of polishing.

@arvindsv -- to answer your questions:

> 1. What happens to someone who was using an older version of the GoCD docker image? Consider cases where they had done a chown 1000:1000 on their shared volume, as well.

Refer [this](https://gist.github.com/GaneshSPatil/dabf81c95b440b77776930bbb4571ecc#3-go-group-migration) 

> 2. What happens to someone using these images with Docker elastic agents? With Kubernetes elastic agents (not in Openshift)?

I doubt if any change is required. 

> 3. What changes need to be done, if any, by anyone using one of these images in the FROM part of a Dockerfile?

Refer [this](https://gist.github.com/GaneshSPatil/dabf81c95b440b77776930bbb4571ecc#gocd-dockerfile-breaking-changes)

---

Apart from the stuff mentioned in the blog post content, we will add a section for:
- Running GoCD as root

**Comment by arvindsv** [comment]
@GaneshSPatil The document you shared look good to me! I guess 19.4.0, 19.5.0 needs to change to 19.5.0, 19.6.0 in it.

**Pull request merged**